content/advent-2016/exposing-go-on-the-internet.md

```diff
@@ -0,0 +1,208 @@
+
+Similarly to `ReadTimeout`, `WriteTimeout` is absolute, with no way to manipulate it from a Handler ([#16100][16100]).
+
+Finally, Go 1.8 [introduces `IdleTimeout`][14204] which limits server-side the amount of time a Keep-Alive connection will be kept idle before being reused. Before Go 1.8, the `ReadTimeout` would start ticking again immediately after a request completed, making it very hostile to Keep-Alive connections: the idle time would consume time the client should have been allowed to send the request, causing unexpected timeouts also for fast clients.
+
+You should set `Read`, `Write` and `Idle` timeouts when dealing with untrusted clients and/or networks, so that a client can't hold up a connection by being slow to write or read.
+
+For detailed background on HTTP/1.1 timeouts (up to Go 1.7) read [my post on the Cloudflare blog][timeouts].
+
+[14204]: https://github.com/golang/go/issues/14204
+[16100]: https://golang.org/issue/16100
+[timeouts]: https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/
+[setreaddeadline]: https://github.com/golang/go/blob/3ba31558d1bca8ae6d2f03209b4cae55381175b3/src/net/http/server.go#L750
+[setwritedeadline]: https://github.com/golang/go/blob/3ba31558d1bca8ae6d2f03209b4cae55381175b3/src/net/http/server.go#L753-L755
+[tlsdeadline]: https://github.com/golang/go/blob/3ba31558d1bca8ae6d2f03209b4cae55381175b3/src/net/http/server.go#L1477-L1483
+
+#### HTTP/2
+
+HTTP/2 is enabled automatically on any Go 1.6+ server if:
+
+* the request is served over TLS/HTTPS
+* `Server.TLSNextProto` is `nil` (setting it to an empty map is how you disable HTTP/2)
+* `Server.TLSConfig` is set and `ListenAndServeTLS` is used  **or**
+* `Serve` is used and `tls.Config.NextProtos` includes `"h2"` (like `[]string{"h2", "http/1.1"}`, since `Serve` is called [too late to auto-modify the TLS Config][15908])
+
+HTTP/2 has a slightly different meaning since the same connection can be serving different requests at the same time, however, they are abstracted to the same set of Server timeouts in Go.
+
+Sadly, `ReadTimeout` breaks HTTP/2 connections in Go 1.7. Instead of being reset for each request it's set once at the beginning of the connection and never reset, breaking all HTTP/2 connections after the `ReadTimeout` duration. [It's fixed in 1.8][16450].
+
+Between this and the inclusion of idle time in `ReadTimeout`, my recommendation is to upgrade to 1.8 as soon as possible.
+
+[15908]: https://github.com/golang/go/issues/15908
+[16450]: https://github.com/golang/go/issues/16450
+
+#### TCP Keep-Alives
+
+If you use `ListenAndServe` (as opposed to passing a `net.Listener` to `Serve`, which offers zero protection by default) a TCP Keep-Alive period of 3 minutes [will be set automatically][tcpKeepAliveListener]. That *will* help with clients that disappear completely off the face of the earth leaving a connection open forever, but I’ve learned not to trust that, and to set timeouts anyway.
```

Comment:
Sadly, I don't have a strong grip on why they don't work. I link to a commit on my Heartbleed project that I had to make because the server was falling over every month, which I checked and was compiled with 1.3, which already had the keep-alives. The following paragraph is my attempt at explaining why. I'll break it into three for clarity.